lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Runtime/HostManager/HostManager.h"
+#include "glow/Backends/DeviceManager.h"
+#include "glow/Graph/Context.h"
+#include "glow/Runtime/RuntimeTypes.h"
+// include Partitioner
+
+#include <future>
+#include <queue>
+
+using namespace glow;
+using namespace runtime;
+
+HostManager::HostManager(std::vector<DeviceConfig> &configs) {
+  DeviceIDTy deviceCount = 0;
+  for (auto config : configs) {
+    auto newDevice = DeviceManager.getDevice(config);
+    devices_.emplace(deviceCount, newDevice);
+    deviceCount++;
+  }
+  provisioner_ = Provisioner();
+}
+
+HostManager::~HostManager() { clearHost(); }
+ResultCode HostManager::addNetwork(Module *M) {
+  std::vector<DeviceInfo> deviceInfo;
+  for (auto device : devices_) {
+    auto info = DeviceInfo();
+    info.availableMemory = device->first->getAvailableMemory();
+  }
+  auto partitioner = Partitioner(M, deviceInfo);
+  auto nodeList = partitioner.partition(M);
+  auto result = provisioner_.provision(nodeList.roots, devices_, M);
+  for (auto node : nodeList.roots) {
+    networks_.emplace(node->name, std::move(node));
+  }
+  for (auto node : nodeList.nodes) {
+    networks_.emplace(node->name, std::move(node));
+  }
+  return result;
+}
+
+void HostManager::removeNetwork(std::string networkID) {
+  // Walk the tree for the given function, calling evict function for each node
+  // before removing it from networks_ which frees the node.
+  std::queue<std::string> nodes;
```

Comment:
Strings are being used because that's the key to the map that stores the nodes. I could use pointers instead and then pointer->name to get the iterator to remove. Using a string seemed cleaner.